pkg/config/validator.go

```diff
@@ -0,0 +1,223 @@
+		strict        bool
+		exprValidator expr.Validator
+		validated     *Validated
+	}
+
+	// Validated store validated configuration.
+	// It has been validated as internally consistent and correct.
+	Validated struct {
+		// Names are given to specific adapter configurations: unique
+		adapterByName map[string]*pb.Adapter
+		// This is more often used to match adapters
+		adapterByKind map[string][]*pb.Adapter
+		serviceConfig *pb.ServiceConfig
+		numAspects    int
+	}
+)
+
+// validateAdapterConfig consumes a yml config string with adapter config.
+// It is validated in presence of validators.
+func (p *Validator) validateGlobalConfig(cfg string) (ce *aspect.ConfigErrors) {
+	m := &pb.GlobalConfig{}
+	err := yaml.Unmarshal([]byte(cfg), m)
+	if err != nil {
+		ce = ce.Append("AdapterConfig", err)
+		return
+	}
+	p.validated.adapterByKind = make(map[string][]*pb.Adapter)
+	p.validated.adapterByName = make(map[string]*pb.Adapter)
+	var acfg proto.Message
+	var aArr []*pb.Adapter
+	var found bool
+	for _, aa := range m.GetAdapters() {
+		if acfg, err = ConvertParams(p.adapterFinder, aa.Impl, aa.Params, p.strict); err != nil {
+			ce = ce.Append("Adapter: "+aa.Impl, err)
+			continue
+		}
+		aa.Params = acfg
+
+		if aArr, found = p.validated.adapterByKind[aa.GetKind()]; !found {
+			aArr = []*pb.Adapter{}
+		}
+		aArr = append(aArr, aa)
+		p.validated.adapterByKind[aa.GetKind()] = aArr
+		if aa.GetName() != "" {
+			p.validated.adapterByName[aa.GetName()] = aa
+		}
+
+	}
+	return
+}
+
+// ValidateSelector ensures that the selector is valid per expression language.
+func (p *Validator) validateSelector(selector string) (err error) {
+	// empty selector always selects
+	if len(selector) == 0 {
+		return nil
+	}
+	return p.exprValidator.Validate(selector)
+}
+
+// validateAspectRules validates the recursive configuration data structure.
+// It is primarily used by validate ServiceConfig.
+func (p *Validator) validateAspectRules(rules []*pb.AspectRule, path string, validatePresence bool) (ce *aspect.ConfigErrors) {
+	var acfg proto.Message
+	var err error
+	for _, rule := range rules {
+		if err = p.validateSelector(rule.GetSelector()); err != nil {
+			ce = ce.Append(path+":Selector "+rule.GetSelector(), err)
+		}
+		path = path + "/" + rule.GetSelector()
+		for idx, aa := range rule.GetAspects() {
+			if acfg, err = ConvertParams(p.managerFinder, aa.GetKind(), aa.GetParams(), p.strict); err != nil {
+				ce = ce.Append(fmt.Sprintf("%s:%s[%d]", path, aa.Kind, idx), err)
+				continue
+			}
+			aa.Params = acfg
+			p.validated.numAspects++
+			if validatePresence {
+				// ensure that aa.Kind has a registered adapter
+				if p.validated.adapterByKind[aa.GetKind()] == nil {
+					ce = ce.Appendf("Kind", "No adapter for Kind=%s is available", aa.GetKind())
+				}
+				if aa.GetAdapter() != "" && p.validated.adapterByName[aa.GetAdapter()] == nil {
+					ce = ce.Appendf("NamedAdapter", "No adapter by name %s Available", aa.GetAdapter())
```

Comment:
if we rephrase the other to match existing suggestion, i suggest we rephrase this to match:

`adapter named '%s' is not available`